mixer/adapter/signalfx/registry.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfx
+
+import (
+	"container/list"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/signalfx/golib/datapoint"
+	"github.com/signalfx/golib/sfxclient"
+)
+
+type metricID string
+
+var currentTime = time.Now
+
+// registry holds a set of collectors that produce datapoints on demand.  This
+// allows us to smooth out the metrics we get to make them regular timeseries.
+// We also need to expire collectors that haven't been updated after a certain
+// amount of time to prevent excessive memory usage.
+type registry struct {
+	cumulatives    map[metricID]*cumulativeCollector
+	rollingBuckets map[metricID]*sfxclient.RollingBucket
+
+	// A linked list that we keep sorted by access time so that we can very
+	// quickly tell which collectors are expired and should be deleted.
+	lastAccessList list.List
+	// A map optimizing lookup of the access time elements that are used in the
+	// above linked list.
+	lastAccesses map[metricID]*list.Element
+
+	expiryTimeout time.Duration
+	lock          sync.RWMutex
```

Comment:
nit: just embed the Mutex ?

```
type registry struct {
   sync.RWMutex
   ...
}
...
   r.Lock()
...
```

Embedding at the top of the struct will also make it clearer just what exactly is being protected by the lock.